File: src/cli-bridge/store.ts
import { randomUUID } from 'node:crypto';
import type { CliBridgeAgent, CliHookResponse } from './types';

export type PendingCliKind = 'permission' | 'question' | 'task_completion';

export interface PendingCliInteraction {
  id: string;
  kind: PendingCliKind;
  source: CliBridgeAgent;
  sessionId: string;
  cwd: string;
  toolName?: string;
  question?: string;
  command?: string;
  hookEventName?: string;
  options?: { label: string; description?: string; preview?: string }[];
  header?: string;
  taskStatus?: 'completed' | 'failed';
  summary?: string;
  replyExpiresAt?: number;
  toolInput?: Record<string, unknown>;
  messageId?: string;
  createdAt: number;
}

const pending = new Map<string, PendingCliInteraction>();
const waiters = new Map<string, (response: CliHookResponse) => void>();
// A decision that arrives before the waiter registers (a fast Feishu click during
// the card POST, before handleMessage reaches waitFor) is buffered here so it is
// not dropped — otherwise waitFor would see the pending already gone and return
// missing_pending, silently losing the user's allow/deny.
const settled = new Map<string, CliHookResponse>();

// Beyond the 24h max IPC/approval wait, anything still in the map leaked: an entry
// created before a sendManagedCard that then threw, or a task-completion card with
// reply continuation off. waitFor's own timeout (≤24h) cleans the happy paths; this
// sweep is the safety net so the map (and findByReply's O(n) scan) can't grow
// unbounded. Lazy on create — no interval lifecycle to manage.
const STALE_PENDING_MS = 25 * 60 * 60_000;

function prunePending(now: number): void {
  for (const [id, item] of pending) {
    if (now - item.createdAt <= STALE_PENDING_MS) continue;
    pending.delete(id);
    waiters.delete(id);
    settled.delete(id);
  }
}

export function createPendingCliInteraction(input: Omit<PendingCliInteraction, 'id' | 'createdAt'>): PendingCliInteraction {
  const now = Date.now();
  prunePending(now);
  const item = { ...input, id: randomUUID(), createdAt: now };
  pending.set(item.id, item);
  return item;
}

export function setPendingCliMessageId(id: string, messageId: string): void {
  const item = pending.get(id);
  if (item) pending.set(id, { ...item, messageId });
}

export function getPendingCliInteraction(id: string): PendingCliInteraction | undefined {
  return pending.get(id);
}

export function findPendingCliInteractionByMessageReply(input: {
  parentId?: string;
  rootId?: string;
}): PendingCliInteraction | undefined {
  const targets = [input.parentId, input.rootId].filter((x): x is string => Boolean(x));
  for (const item of pending.values()) {
    if (item.messageId && targets.includes(item.messageId)) return item;
  }
  return undefined;
}

export function resolvePendingCliInteraction(id: string, response: CliHookResponse): boolean {
  const item = pending.get(id);
  if (!item) return false;
  pending.delete(id);
  const waiter = waiters.get(id);
  if (waiter) {
    waiters.delete(id);
    waiter(response);
  } else {
    // No waiter yet (resolve raced ahead of waitFor) — buffer for it to pick up.
    settled.set(id, response);
  }
  return true;
}

export function waitForPendingCliInteraction(id: string, timeoutMs: number): Promise<CliHookResponse> {
  return new Promise((resolve) => {
    // A decision may have already arrived before we got here (see `settled`).
    const buffered = settled.get(id);
    if (buffered) {
      settled.delete(id);
      resolve(buffered);
      return;
    }
    if (!pending.has(id)) {
      resolve({ decision: 'fallback_local', reason: 'missing_pending' });
      return;
    }
    const timer = setTimeout(() => {
      pending.delete(id);
      waiters.delete(id);
      settled.delete(id);
      resolve({ decision: 'fallback_local', reason: 'timeout' });
    }, Math.max(timeoutMs, 0));
    waiters.set(id, (response) => {
      clearTimeout(timer);
      resolve(response);
    });
  });
}

File: internal/clibridge/store.go
package clibridge

// store.go —— 进程内 pending interaction 仓库（对齐 TS cli-bridge/store.ts）。
// 一个 hook 卡 = 一条 pending；飞书点击 / 本机回归 / 超时都会 resolve 它，
// 唤醒阻塞在对应 waitFor 上的 IPC 连接（即等审批的 agent hook）。
//
// 关键不变量：
//   - settled 缓存「先到的决策」：飞书点击快过 waitFor 注册时，决策先入 settled，
//     waitFor 起来直接取走，不丢点击。
//   - STALE_PENDING_MS 兜底清扫，map（及 findByReply 的 O(n) 扫描）不会无限增长。

import (
	"crypto/rand"
	"encoding/hex"
	"sync"
	"time"
)

// PendingCliKind pending 类型。
type PendingCliKind string

const (
	PendingPermission      PendingCliKind = "permission"
	PendingQuestion        PendingCliKind = "question"
	PendingTaskCompletion  PendingCliKind = "task_completion"
)

// PendingCliInteraction 一条 pending（审批 / 问答 / 完成）。
type PendingCliInteraction struct {
	ID           string            `json:"id"`
	Kind         PendingCliKind    `json:"kind"`
	Source       CliBridgeAgent    `json:"source"`
	SessionID    string            `json:"sessionId"`
	Cwd          string            `json:"cwd"`
	ToolName     string            `json:"toolName,omitempty"`
	Question     string            `json:"question,omitempty"`
	Command      string            `json:"command,omitempty"`
	HookEventName string           `json:"hookEventName,omitempty"`
	Options      []CliQuestionOption `json:"options,omitempty"`
	Header       string            `json:"header,omitempty"`
	// AskUserQuestion：整组 1-4 问题（一张多问题表单）。
	Questions    []CliQuestionItem `json:"questions,omitempty"`
	TaskStatus   string            `json:"taskStatus,omitempty"` // completed|failed
	Summary      string            `json:"summary,omitempty"`
	ReplyExpiresAt int64           `json:"replyExpiresAt,omitempty"`
	ToolInput    map[string]any    `json:"toolInput,omitempty"`
	MessageID    string            `json:"messageId,omitempty"`
	CreatedAt    int64             `json:"createdAt"`
}

// 超过 24h 上界（IPC/审批等待上限）仍滞留的 pending 即泄漏（send 抛错未建、或
// 完成卡 reply 关掉），此处统一清扫。懒清除，无定时生命周期。
const stalePendingMs = 25 * 60 * 60 * 1000

type pendingStore struct {
	mu       sync.Mutex
	pending  map[string]*PendingCliInteraction
	waiters  map[string]chan CliHookResponse
	settled  map[string]CliHookResponse
}

var store = &pendingStore{
	pending: map[string]*PendingCliInteraction{},
	waiters: map[string]chan CliHookResponse{},
	settled: map[string]CliHookResponse{},
}

func uuid() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		// 极不可能；退化为时间+随机。
		return time.Now().Format("20060102150405.000000000")
	}
	return hex.EncodeToString(b)
}

// prunePending 清除过期 pending（含其 waiter/settled）。
func (s *pendingStore) prunePending(now int64) {
	for id, item := range s.pending {
		if now-item.CreatedAt <= stalePendingMs {
			continue
		}
		delete(s.pending, id)
		delete(s.waiters, id)
		delete(s.settled, id)
	}
}

// CreatePendingCliInteraction 新建一条 pending 并写入仓库。
func CreatePendingCliInteraction(input PendingCliInteraction) *PendingCliInteraction {
	now := time.Now().UnixMilli()
	store.mu.Lock()
	store.prunePending(now)
	item := input
	item.ID = uuid()
	item.CreatedAt = now
	cp := item
	store.pending[item.ID] = &cp
	store.mu.Unlock()
	return &cp
}

// SetPendingCliMessageId 回填飞书卡片 message_id（resolve 时按回复匹配用）。
func SetPendingCliMessageId(id, messageID string) {
	store.mu.Lock()
	if item, ok := store.pending[id]; ok {
		cp := *item
		cp.MessageID = messageID
		store.pending[id] = &cp
	}
	store.mu.Unlock()
}

// GetPendingCliInteraction 取一条 pending（只读）。
func GetPendingCliInteraction(id string) *PendingCliInteraction {
	store.mu.Lock()
	defer store.mu.Unlock()
	if item, ok := store.pending[id]; ok {
		cp := *item
		return &cp
	}
	return nil
}

// FindPendingCliInteractionByMessageReply 按回复的 parent/root 匹配 pending。
func FindPendingCliInteractionByMessageReply(input struct {
	ParentID string
	RootID   string
}) *PendingCliInteraction {
	targets := make([]string, 0, 2)
	if input.ParentID != "" {
		targets = append(targets, input.ParentID)
	}
	if input.RootID != "" {
		targets = append(targets, input.RootID)
	}
	store.mu.Lock()
	defer store.mu.Unlock()
	for _, item := range store.pending {
		if item.MessageID != "" {
			for _, t := range targets {
				if item.MessageID == t {
					cp := *item
					return &cp
				}
			}
		}
	}
	return nil
}

// ResolvePendingCliInteraction 决策到达：唤醒 waiter 或缓冲到 settled。
func ResolvePendingCliInteraction(id string, response CliHookResponse) bool {
	store.mu.Lock()
	if _, ok := store.pending[id]; !ok {
		store.mu.Unlock()
		return false
	}
	delete(store.pending, id)
	waiter, hasWaiter := store.waiters[id]
	if hasWaiter {
		delete(store.waiters, id)
		store.mu.Unlock()
		waiter <- response
		return true
	}
	store.settled[id] = response
	store.mu.Unlock()
	return true
}

// WaitForPendingCliInteraction 阻塞等决策（≤ timeoutMs）。先查 settled 缓冲。
// 出错路径：settled 有值直接取；pending 已无则 missing_pending；超时 timeout。
func WaitForPendingCliInteraction(id string, timeoutMs int) CliHookResponse {
	store.mu.Lock()
	if buffered, ok := store.settled[id]; ok {
		delete(store.settled, id)
		store.mu.Unlock()
		return buffered
	}
	if _, ok := store.pending[id]; !ok {
		store.mu.Unlock()
		return CliHookResponse{Decision: DecisionFallbackLocal, Reason: "missing_pending"}
	}
	ch := make(chan CliHookResponse, 1)
	store.waiters[id] = ch
	store.mu.Unlock()

	timer := time.NewTimer(time.Duration(max(timeoutMs, 0)) * time.Millisecond)
	defer timer.Stop()
	select {
	case resp := <-ch:
		return resp
	case <-timer.C:
		store.mu.Lock()
		delete(store.pending, id)
		delete(store.waiters, id)
		delete(store.settled, id)
		store.mu.Unlock()
		return CliHookResponse{Decision: DecisionFallbackLocal, Reason: "timeout"}
	}
}
